lara-typescript/src/interactive-api-parent/iframe-saver.ts

```diff
@@ -278,6 +261,19 @@ export class IFrameSaver {
       return this.getFirebaseJwt(request);
     });
 
+    // add listeners from "plugins"
+    this.plugins.forEach(plugin => {
+      if (plugin.listeners) {
+        for (const m in plugin.listeners) {
+          if (Object.prototype.hasOwnProperty.call(plugin.listeners, m)) {
+            const msg = m as LaraInteractiveApi.IFrameSaverClientMessage;
+            const listener = msg && plugin.listeners[msg];
+            listener && this.addListener(msg, listener);
+          }
+        }
+      }
+    });
+
```

Comment:
Some plugins will want to send messages and they might also want to remove their listener.
So I'd think it would be better to just initialize the plugins with the iframePhone object and then let them add their own listeners.

However one thing to watch out for (I think) is that iframePhone doesn't allow more than one listener for a message at at time. So one plugin could mess up another.

The interactive api client has a iframephone manager of some kind that manages this better so there can multiple listeners for a message.

Reply:
To my mind, the goal of this effort is purely one of implementation -- make it possible to separate code so that `iframe-saver.ts` doesn't become any more unmaintainable than it already is. This implementation doesn't introduce any additional problems/constraints that weren't already present in the existing inline implementation. Building a robust plugin API here seems beyond the scope of this PR.

That said, I don't have a problem with passing the IFramePhoneManager into the plugin if you'd prefer (I made plugins functions so that there could be arguments) and having the plugin add the listeners itself, but issues like multiple plugins adding multiple listeners for the same message should be treated the way they're treated now -- as programmer error.

If we want to flesh this out into something more robust than simply a technique for separating implementation into multiple files, then we can design such a plugin API and write stories for it.

Reply:
Right, it doesn't need to be a full plugin system. I'm just anticipating that the very next plugin will need to send messages as well as receive them. So then reducing the amount of layers that we need to maintain makes this even more simple.

Is frame-saver already using a IFramePhoneManager? It looks like it is just using IFramePhone directly.

Reply:
It's just using the `ParentEndpoint` from iFramePhone directly. I just pushed a commit which changes the plugin to take iFramePhone as an argument. It took a while to get the tests stubbed/mocked properly but I think I've wrestled it to the ground. It makes the `post()` method available to plugins.